File: go/0347.top-k-frequent-elements/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:34
// leetgo: dev
// https://leetcode.com/problems/top-k-frequent-elements/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

// Bucket-sort approach (faster) — used as the main solution.
func topKFrequent(nums []int, k int) []int {
	hash := make(map[int]int, k)
	for _, val := range nums {
		hash[val] += 1
	}

	count := make([][]int, len(nums)+1)
	for key, val := range hash {
		count[val] = append(count[val], key)
	}

	final := []int{}
	for i := len(count) - 1; i >= 0; i-- {
		for j := 0; j < len(count[i]); j++ {
			final = append(final, count[i][j])
			if len(final) == k {
				return final
			}
		}
	}
	return nil
}

// First approach (slower) — kept for reference/comparison.
func topKFrequentNaive(nums []int, k int) []int {
	hash := make(map[int]int, k)
	maxx := 0

	for _, val := range nums {
		hash[val] += 1
		maxx = max(maxx, hash[val])
	}

	array := []int{}
	for i := 0; i < k; i++ {
		for key, val := range hash {
			if val == maxx {
				array = append(array, key)
				delete(hash, key)
				maxx = 0
			}
		}
		for _, val := range hash {
			maxx = max(maxx, val)
		}
	}
	return array
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	k := Deserialize[int](ReadLine(stdin))
	ans := topKFrequent(nums, k)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0347.top-k-frequent-elements/solution.rs
// Created by Olgierd Palasz at 2026/07/15 21:18
// leetgo: dev
// https://leetcode.com/problems/top-k-frequent-elements/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::collections::{BinaryHeap, HashMap};
impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut hashed: HashMap<i32, usize> = HashMap::new();

        for num in nums.into_iter() {
            *hashed.entry(num).or_insert(0) += 1;
        }

        let mut heap: BinaryHeap<(usize, i32)> =
            hashed.into_iter().map(|(val, cnt)| (cnt, val)).collect();

        let mut result: Vec<i32> = Vec::with_capacity(k as usize);

        for _ in 0..k {
            result.push(heap.pop().map(|(_, value)| value).unwrap());
        }

        result
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let k: i32 = deserialize(&read_line()?)?;
    let ans: Vec<i32> = Solution::top_k_frequent(nums, k);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
